src/tiled/mapobjectitem.cpp

```diff
@@ -140,6 +140,12 @@ void MapObjectItem::paint(QPainter *painter,
     const auto renderer = mMapDocument->renderer();
     const qreal painterScale = renderer->painterScale();
     const QColor color = mIsHoveredIndicator ? mColor.lighter() : mColor;
+    QColor fillColor = color;
+    fillColor.setAlpha(50);
+    if (auto type = Object::propertyTypes().findClassFor(mapObject()->className(), *mapObject()))
+        if (!type->drawFill)
+            fillColor.setAlpha(0);
```

Comment:
Of course, we're duplicating the logic for the fill color all over the place like this. I think it would make sense to integrate it into `MapObject::effectiveColor`, which could either be changed to take two reference parameters, like:
```c++
void MapObject::effectiveColors(QColor &color, QColor &fillColor) const
```
Or we could return a structure:
```c++
struct MapObjectColors
{
    QColor main;
    QColor fill;
};

MapObjectColors MapObject::effectiveColors() const
```
I think I'd prefer the latter, and we can keep `effectiveColor` as convenience since many places in the code only need the main color.

Also, we can add this struct as member of `MapObjectItem` to replace the current `QColor` member, so we can also check when the fill colors changes in `syncWithMapObject` and not look up the class each time an object is rendered.